### PR TITLE
Add KafkaProducerProcess

### PR DESCRIPTION
**Pull request by cpx86**
As discussed here and there before, `Process` can be used to abstract things like message queues behind a `PID`.

I've added an initial stab at a process that wraps a Kafka producer. It doesn't currently have any way to customize the serializer, so ideas ot fix this would be welcome :)

### CODE REVIEW

**Comment by tomliversidge** [comment]
I missed discussions around this, so could you explain what's the difference between implementing this at the process level rather than as an actor? I.e. Do we lose supervision capabilities this way?

File: src/Proto.Actor/ProcessRegistry.cs [hunk 1]
```diff
@@ -47,12 +47,24 @@ public Process Get(PID pid)
...
+            };
+            var ok = _localActorRefs.TryAdd(pid.Id, aref);
+            return (pid, ok);
+        }
+
         public (PID pid, bool ok) TryAdd(string id, Process aref)
         {
             var pid = new PID
             {
                 Id = id,
```
Comment (adamhathcock) [inline src/Proto.Actor/ProcessRegistry.cs]: Rather than have the above additional method can you just do: `Id = id ?? NextId();` ?

Comment (adamhathcock) [inline src/Proto.Actor/ProcessRegistry.cs]: I guess you like having the overload that takes no Id though.  Just trying to collapse duplicated code.

**Comment by raskolnikoov** [comment]
I've built many systems using kafka and actors together and it works great. So this is basically a plugin implementation with the confluent api? Looks good but the producer needs to have more config options. And maybe we should have a ConsumerProcess also that reports back all the messages it consumes from kafka to its parent actor? Thats how i currently do it in my daily work applications.

**Comment by rogeralsing** [comment]
Me and @cpx86 have been discussing this for quite a while, that Actors are just one form of message passing construct. there can be others, Kafka, Rabbit, SignalR etc.
All of those concepts could be given a PID so that you can send messages to that PID.
(Just like Routers do, as they have a special process which ignores the actor concurrency constraint)

This can be implemented as a different kind of process as done here, or it could be implemented using actors that then delegate the messages forward to the underlying construct.

Just brainstorm away on what would be the proper way forward.
Custom Process = We can ignore actor concurrency and mailbox.
Wrapper Actor = We can leverage supervision and lifecycles.

Go :-)

File: src/Proto.Actor/ProcessRegistry.cs (hunk 1 above)
Comment (cpx86) [inline src/Proto.Actor/ProcessRegistry.cs]: Ah maybe the first method should call the second method?

**Comment by cpx86** [comment]
One downside I noticed with the current PID/Process signatures is that they're not async. Would it make sense to have them return a Task?

**Comment by rogeralsing** [comment]
@cpx86 agree, Tell and the downstream methods including mailbox methods really should have been async.

**Comment by cpx86** [comment]
@rogeralsing Should we change it? Shouldn't break too much.

**Comment by cpx86** [comment]
Added another example that I think highlights the usefulness of this a bit better: adding an actor and a KafkaProducerProcess to a broadcast router, so all messages are sent both to an in-memory actor and persisted onto a Kafka topic.

**Comment by cpx86** [comment]
@raskolnikoov yeah I was thinking of adding a consumer process as well, but I'm not sure how it would look. Any suggestions? Maybe something like the EventStream?

**Comment by cpx86** [comment]
On second thought, let's not go to Camelot. 'Tis a silly place.

**Pull request closed without merging**